email-config-shell/src/driver.py

```diff
@@ -0,0 +1,140 @@
+from cloudshell.shell.core.resource_driver_interface import ResourceDriverInterface
+from cloudshell.shell.core.driver_context import InitCommandContext, ResourceCommandContext, AutoLoadResource, \
+    AutoLoadAttribute, AutoLoadDetails, CancellationContext
+from cloudshell.shell.core.session.logging_session import LoggingSessionContext
+from cloudshell.shell.core.session.cloudshell_session import CloudShellSessionContext
+from cloudshell.core.context.error_handling_context import ErrorHandlingContext
+from cloudshell.email import EmailConfig, EmailService
+#from data_model import *  # run 'shellfoundry generate' to generate data model classes
+
+
+class EmailConfigShellDriver (ResourceDriverInterface):
+
+    def __init__(self):
+        """
+        ctor must be without arguments, it is created with reflection at run time
+        """
+        pass
+
+    def initialize(self, context):
+        """
+        Initialize the driver session, this function is called everytime a new instance of the driver is created
+        This is a good place to load and cache the driver configuration, initiate sessions etc.
+        :param InitCommandContext context: the context the command runs on
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Destroy the driver session, this function is called everytime a driver instance is destroyed
+        This is a good place to close any open sessions, finish writing to log files
+        """
+        pass
+
+    # <editor-fold desc="Discovery">
+
+    def get_inventory(self, context):
+        """
+        Discovers the resource structure and attributes.
+        :param AutoLoadCommandContext context: the context the command runs on
+        :return Attribute and sub-resource information for the Shell resource you can return an AutoLoadDetails object
+        :rtype: AutoLoadDetails
+        """
+
+        with LoggingSessionContext(context) as logger:
+            with ErrorHandlingContext(logger):
+                with CloudShellSessionContext(context) as session:
+                    decrypted_password = session.DecryptPassword(
+                                   context.resource.attributes['Email Config Shell.Password']).Value
+                    emailconfig = EmailConfig(
+                        context.resource.attributes.get('Email Config Shell.SMTP Server'),
+                        context.resource.attributes.get('Email Config Shell.User'),
+                        decrypted_password,
+                        context.resource.attributes.get('Email Config Shell.From Address'),
+                        context.resource.attributes.get('Email Config Shell.SMTP Port')
+                    )
+                    emailservice = EmailService(emailconfig, logger)
+                    emailservice.validate_email_config()
+
+        return AutoLoadDetails([], [])
```

Comment:
What's written here looks 100% correct. I do have a couple of small suggestions for improvements in terms of code readability:

1. several "with" statements can be grouped in the same line - benefic is to have less indentations
2. using an object to access the shell attributes instead of hard coding strings

Code example:

        with LoggingSessionContext(context) as logger, ErrorHandlingContext(logger):
            with CloudShellSessionContext(context) as session:
                email_config_shell = EmailConfigShell.create_from_context(context)
                decrypted_password = session.DecryptPassword(email_config_shell.password).Value
                emailconfig = EmailConfig(
                    email_config_shell.smtp_server,
                    email_config_shell.user,
                    decrypted_password,
                    email_config_shell.from_address,
                    email_config_shell.smtp_port
                )
                emailservice = EmailService(emailconfig, logger)
                emailservice.validate_email_config()

        return AutoLoadDetails([], [])

Reply:
Sound good! Done.